tokio/src/util/linked_list.rs

```diff
@@ -313,6 +347,146 @@ cfg_taskdump! {
     }
 }
 
+// ===== impl AtomicLinkedList =====
+
+feature! {
+    #![feature = "sync"]
+
+    use core::sync::atomic::{
+        AtomicPtr,
+        Ordering::{AcqRel, Relaxed},
+    };
+
+    /// An atomic intrusive linked list. It allows pushing new nodes concurrently.
+    /// Removing nodes still requires an exclusive reference.
+    pub(crate) struct AtomicLinkedList<L, T> {
+        /// Linked list head.
+        head: AtomicPtr<T>,
+
+        /// Linked list tail.
+        tail: UnsafeCell<Option<NonNull<T>>>,
+
+        /// Node type marker.
+        _marker: PhantomData<*const L>,
+    }
+
+    unsafe impl<L: Link> Send for AtomicLinkedList<L, L::Target> where L::Target: Send {}
+    unsafe impl<L: Link> Sync for AtomicLinkedList<L, L::Target> where L::Target: Sync {}
+
+    impl<L: Link> Default for AtomicLinkedList<L, L::Target> {
+        fn default() -> Self {
+            Self::new()
+        }
+    }
+
+    impl<L, T> AtomicLinkedList<L, T> {
+        /// Creates an empty atomic linked list.
+        pub(crate) const fn new() -> AtomicLinkedList<L, T> {
+            AtomicLinkedList {
+                head: AtomicPtr::new(core::ptr::null_mut()),
+                tail: UnsafeCell::new(None),
+                _marker: PhantomData,
+            }
+        }
+
+        /// Convert an atomic linked list into a non-atomic version.
+        pub(crate) fn into_list(mut self) -> LinkedList<L, T> {
+            LinkedList {
+                head: NonNull::new(*self.head.get_mut()),
+                tail: *self.tail.get_mut(),
+                _marker: PhantomData,
+            }
+        }
+    }
+
+    impl<L: Link> LinkedListBase<L> for AtomicLinkedList<L, L::Target> {
+        fn head(&mut self) -> Option<NonNull<L::Target>> {
+            NonNull::new(*self.head.get_mut())
+        }
+
+        fn tail(&mut self) -> Option<NonNull<L::Target>> {
+            *self.tail.get_mut()
+        }
+
+        fn set_head(&mut self, node: Option<NonNull<L::Target>>) {
+            *self.head.get_mut() = match node {
+                Some(ptr) => ptr.as_ptr(),
+                None => core::ptr::null_mut(),
+            };
+        }
+
+        fn set_tail(&mut self, node: Option<NonNull<L::Target>>) {
+            *self.tail.get_mut() = node;
+        }
+    }
+
+    impl<L: Link> AtomicLinkedList<L, L::Target> {
+        /// Atomically adds an element first in the list.
+        /// This method can be called concurrently from multiple threads.
+        ///
+        /// # Safety
+        ///
+        /// The caller must ensure that:
+        /// - `val` is not pushed concurrently by muptiple threads,
+        /// - `val` is not already part of some list.
+        pub(crate) unsafe fn push_front(&self, val: L::Handle) {
```

Comment:
This is an interesting implementation. It looks like it could be correct. I'll have to think more about it.